flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sinks/QueryableTableSink.scala

```diff
@@ -0,0 +1,223 @@
+      keys,
+      getFieldNames,
+      getFieldTypes,
+      calculateCleanupTimeDomain(dataStream.getExecutionEnvironment.getStreamTimeCharacteristic))
+
+    dataStream.keyBy(new RowKeySelector(keyIndices, keySelectorType))
+      .process(processFunction)
+  }
+
+  private def calculateCleanupTimeDomain(timeCharacteristic: TimeCharacteristic): TimeDomain = {
+    val timeDomainFromTimeCharacteristic = {
+      timeCharacteristic match {
+        case TimeCharacteristic.IngestionTime | TimeCharacteristic.ProcessingTime =>
+          TimeDomain.PROCESSING_TIME
+        case TimeCharacteristic.EventTime =>
+          TimeDomain.EVENT_TIME
+      }
+    }
+
+    cleanupTimeDomain.getOrElse(timeDomainFromTimeCharacteristic)
+  }
+
+  override protected def copy: TableSinkBase[JTuple2[JBool, Row]] = {
+    new QueryableTableSink(this.namePrefix, this.queryConfig, this.cleanupTimeDomain)
+  }
+}
+
+class RowKeySelector(
+  private val keyIndices: Array[Int],
+  @transient private val returnType: TypeInformation[Row])
+  extends KeySelector[JTuple2[JBool, Row], Row]
+    with ResultTypeQueryable[Row] {
+
+  override def getKey(value: JTuple2[JBool, Row]): Row = {
+    val keys = keyIndices
+
+    val srcRow = value.f1
+
+    val destRow = new Row(keys.length)
+    var i = 0
+    while (i < keys.length) {
+      destRow.setField(i, srcRow.getField(keys(i)))
+      i += 1
+    }
+
+    destRow
+  }
+
+  override def getProducedType: TypeInformation[Row] = returnType
+}
+
+class QueryableStateProcessFunction(
+    private val namePrefix: String,
+    private val queryConfig: StreamQueryConfig,
+    private val keyNames: Array[String],
+    private val fieldNames: Array[String],
+    private val fieldTypes: Array[TypeInformation[_]],
+    private val cleanupTimeDomain: TimeDomain)
+  extends KeyedProcessFunctionWithCleanupState[Row, JTuple2[JBool, Row], Void](queryConfig) {
+
+  @transient private var states = Array[ValueState[AnyRef]]()
+  @transient private var nonKeyIndices = Array[Int]()
+
+  override def open(parameters: Configuration): Unit = {
+    super.open(parameters)
+
+    nonKeyIndices = fieldNames.indices
+      .filter(idx => !keyNames.contains(fieldNames(idx)))
+      .toArray
+
+    val statesBuilder = Array.newBuilder[ValueState[AnyRef]]
+
+    for (i <- nonKeyIndices) {
+      val stateDesc = new ValueStateDescriptor(fieldNames(i), fieldTypes(i))
+      stateDesc.initializeSerializerUnlessSet(getRuntimeContext.getExecutionConfig)
+      stateDesc.setQueryable(s"$namePrefix-${fieldNames(i)}")
+      statesBuilder += getRuntimeContext.getState(stateDesc).asInstanceOf[ValueState[AnyRef]]
+    }
+
+    states = statesBuilder.result()
+
+    initCleanupTimeState("QueryableStateCleanupTime")
+  }
+
+  override def processElement(
+    value: JTuple2[JBool, Row],
+    ctx: KeyedProcessFunction[Row, JTuple2[JBool, Row], Void]#Context,
+    out: Collector[Void]): Unit = {
+    if (value.f0) {
+      for (i <- nonKeyIndices.indices) {
```

Comment:
We don't use Scala magic in runtime functions for performance reasons. Use a basic while loop instead.